include/boost/beast/experimental/core/impl/url_parser.ipp

```diff
@@ -0,0 +1,352 @@
+//
+// Copyright (c) 2018 jackarain (jack dot wgm at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#include <boost/beast/experimental/core/url_parser.hpp>
+
+namespace boost {
+namespace beast {
+
+//--------------------------------------------------------------------------
+// scheme:[//[user[:password]@]host[:port]][/path][?query][#fragment]
+
+using boost::beast::string_view;
+
+string_view url_parser::make_string_view(const char* b, std::size_t s)
+{
+    return string_view(b, s);
+}
+
+bool url_parser::isunreserved(const char c) const
+{
+    if (std::isalpha(c) || std::isdigit(c) ||
+        c == '-' || c == '.' || c == '_' || c == '~')
+        return true;
+    return false;
+}
+
+bool url_parser::isuchar(const char c) const
+{
+    if (isunreserved(c) || c == ';' || c == '?' || c == '&' || c == '=')
+        return true;
+    return false;
+}
+
+bool url_parser::ishsegment(const char c) const
+{
+    if (isuchar(c) ||
+        c == ';' || c == ':' || c == '@' || c == '&' || c == '=')
+        return true;
+    return false;
+}
+
+bool url_parser::issubdelims(const char c) const
+{
+    if (c == '!' || c == '$' || c == '\'' || c == '(' ||
+        c == ')' || c == '*' || c == '+' || c == ',' ||
+        c == '=')
+        return true;
+    return false;
+}
+
+//--------------------------------------------------------------------------
+
+bool url_parser::parse(const std::string& url)
+{
+    enum
+    {
+        scheme_start,
+        scheme,
+        slash_start,
+        slash,
+        username_start,
+        username,
+        password_start,
+        password,
+        host_start,
+        host,
+        port_start,
+        port,
+        path,
+        query,
+        fragment
+    } state = scheme_start;
+
+    auto b = url.c_str();
+    auto e = url.c_str() + url.size();
+    auto part_start = b;
+    bool is_ipv6 = false;
+
+    while (b != e)
+    {
+        auto c = *b++;
+        switch (state)
+        {
+        case scheme_start:
+            if (std::isalpha(c))
+            {
+                state = scheme;
+                continue;
+            }
+            return false;
+        case scheme:
+            if (std::isalpha(c) || std::isdigit(c) ||
```

Comment:
Does the scheme allow uppercase? I'm not sure. If it does, we probably need to convert it to lower case.

Reply:
Will check the RFC, and convert if it allows uppercase.